lib/shopify_app/admin_api/with_token_refetch.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module ShopifyApp
+  module AdminAPI
+    module WithTokenRefetch
+      def with_token_refetch(session, session_token)
+        retrying = false if retrying.nil?
+        yield
+      rescue ShopifyAPI::Errors::HttpResponseError => error
+        if error.code == 401 && !retrying
+          retrying = true
+          ShopifyApp::Logger.debug("Encountered 401 error, exchanging token and retrying with new access token")
+          new_session = ShopifyApp::Auth::TokenExchange.perform(session_token)
+          copy_session_attributes(from: new_session, to: session)
```

Comment:
Should we just reassign `session = new_session` instead, and completely override it?

Reply:
We need to update the same instance because of possible subsequent calls in the same request that might be using it. Performing token exchange saves the new session to the db but in memory we still hold the original values.

An example would be to fetch a product, then updating it, performing 2 calls. If the app does this in separate places with separate `with_token_refetch` calls, it would end up causing token exchange to happen twice, because `current_active_session` (or even just their `session` variable) would still contain the old access token with its old expiry, scopes etc when it performs the second API call.